lambda/src/server/slack_db_linking.cljs

```diff
@@ -0,0 +1,151 @@
+(ns server.slack-db-linking
+  (:require [server.tokens :as tokens]
+            [server.firebase :as fire]
+            [kitchen-async.promise :as p]
+            [server.slack :as slack]
+            [cljs.pprint :as pp]
+            [applied-science.js-interop :as j]))
+
+;; LINKING
+;;
+;; Firebase schema -
+;; /slack-channel/{id}   => {project-id, workspace-id}
+;; /slack-user/{id}      => {account-id, workspace-id}
+;; /slack-workspace/{id} => {board-id}
+
+(defn slack-user-by-email
+  "Returns slack user id for the given email address, if found"
+  [email]
+  (p/let [res (slack/get+ "users.lookupByEmail" {:query {:email email}})]
+    ;; UNCLEAR: how does Slack know what workspace we are interested in?
+    (when (j/get res :ok)
+      ;; user contains team_id, id, is_admin
+      (j/get res :user))))
+
+;; TEAM ID
+;; team - the team ID belonging to a target workspace. Useful when you know which
+;; workspace the user should be sent to. When no team is provided or the user is not
+;; signed in, users will be sent to their default team or asked to sign in. Use the
+;; team.info method to obtain the ID of the workspace your app is currently installed in.
+
+;; Create link entries
+
+(defn link-workspace-to-board!
+  [workspace-id board-id]
+  (fire/put+ (str "/slack-workspace/" workspace-id)
+             {:body {:board-id board-id}}))
+
+(defn link-channel-to-project!
+  [workspace-id channel-id project-id]
+  (fire/put+ (str "/slack-channel/" channel-id)
+             {:body {:workspace-id workspace-id
+                     :project-id project-id}}))
+
+(defn link-user-to-account!
+  [workspace-id user-id account-id]
+  (fire/put+ (str "/slack-user/" user-id)
+             {:body {:workspace-id workspace-id
+                     :account-id account-id}}))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Read link entries
+
+;; direct lookups
+
+(defn linked-channel [channel-id]
+  (fire/get+ (str "/slack-channel/" channel-id)))
+
+(defn linked-user [user-id]
+  (fire/get+ (str "/slack-user/" user-id)))
+
+(defn linked-workspace [workspace-id]
+  (fire/get+ (str "/slack-workspace/" workspace-id)))
+
+;; lookups by index
+
+(defn workspace->all-linked-channels [workspace-id]
+  (p/->> (fire/get+ (str "/slack-channel")
+                    {:query {:orderBy "workspace-id"
+                             :equalTo workspace-id}})
+         (fire/obj->list :channel-id)))
+
+(defn project->channel [project-id]
+  (p/->> (fire/get+ (str "/slack-channel")
+                    {:query {:orderBy "project-id"
+                             :equalTo project-id
+                             :limitToFirst 1}})
+         (fire/obj->list :channel-id)))
+
+(defn account->all-slack-users [account-id]
```

Comment:
I'm not sure about this, because it may get easier as I grow more familiar, but it still takes me a second to mentally look up whether `account`, `project`, `workspace` is a Slack or SparkBoard entity. We may want to consider annotating this with a comment, prefix, or suffix.